internal/querynode/shard_cluster.go

```diff
@@ -0,0 +1,477 @@
+		return
+	}
+
+	if old.nodeID != evt.nodeID {
+		log.Warn("ShardCluster removeSegment found node not match", zap.Int64("segmentID", evt.segmentID), zap.Int64("nodeID", old.nodeID), zap.Int64("evtNodeID", evt.nodeID))
+		return
+	}
+
+	//TODO check handoff / load balance
+	delete(sc.segments, evt.segmentID)
+}
+
+// init list all nodes and semgent states ant start watching
+func (sc *ShardCluster) init() {
+	// list nodes
+	nodes, nodeEvtCh := sc.nodeDetector.watchNodes(sc.collectionID, sc.replicaID, sc.vchannelName)
+	for _, node := range nodes {
+		sc.addNode(node)
+	}
+	go sc.watchNodes(nodeEvtCh)
+
+	// list segments
+	segments, segmentEvtCh := sc.segmentDetector.watchSegments(sc.collectionID, sc.replicaID, sc.vchannelName)
+	for _, segment := range segments {
+		sc.updateSegment(segment)
+	}
+	go sc.watchSegments(segmentEvtCh)
+
+	sc.healthCheck()
+}
+
+// healthCheck iterate all segments to to check cluster could provide service.
+func (sc *ShardCluster) healthCheck() {
+	for _, segment := range sc.segments {
+		if segment.state != segmentStateLoaded { // TODO check hand-off or load balance
+			sc.state.Store(int32(unavailable))
+			return
+		}
+	}
+	sc.state.Store(int32(available))
+}
+
+// watchNodes handles node events.
+func (sc *ShardCluster) watchNodes(evtCh <-chan nodeEvent) {
+	for {
+		select {
+		case evt, ok := <-evtCh:
+			if !ok {
+				log.Warn("ShardCluster node channel closed", zap.Int64("collectionID", sc.collectionID), zap.Int64("replicaID", sc.replicaID))
+				return
+			}
+			switch evt.eventType {
+			case nodeAdd:
+				sc.addNode(evt)
+			case nodeDel:
+				sc.removeNode(evt)
+			}
+		case <-sc.closeCh:
+			log.Info("ShardCluster watchNode quit", zap.Int64("collectionID", sc.collectionID), zap.Int64("replicaID", sc.replicaID), zap.String("vchannelName", sc.vchannelName))
+			return
+		}
+	}
+}
+
+// watchSegments handles segment events.
+func (sc *ShardCluster) watchSegments(evtCh <-chan segmentEvent) {
+	for {
+		select {
+		case evt, ok := <-evtCh:
+			if !ok {
+				log.Warn("ShardCluster segment channel closed", zap.Int64("collectionID", sc.collectionID), zap.Int64("replicaID", sc.replicaID))
+				return
+			}
+			switch evt.eventType {
+			case segmentAdd:
+				sc.updateSegment(evt)
+			case segmentDel:
+				sc.removeSegment(evt)
+			}
+		case <-sc.closeCh:
+			log.Info("ShardCluster watchSegments quit", zap.Int64("collectionID", sc.collectionID), zap.Int64("replicaID", sc.replicaID), zap.String("vchannelName", sc.vchannelName))
+			return
+		}
+	}
+}
+
+func (sc *ShardCluster) getNode(nodeID int64) (*shardNode, bool) {
+	sc.mut.RLock()
+	defer sc.mut.RUnlock()
+	node, ok := sc.nodes[nodeID]
+	return node, ok
+}
+
+func (sc *ShardCluster) getSegment(segmentID int64) (*shardSegmentInfo, bool) {
+	sc.mut.RLock()
+	defer sc.mut.RUnlock()
+	segment, ok := sc.segments[segmentID]
+	return segment, ok
+}
+
+// segmentAllocations returns segment to node mappings.
```

Comment:
I guess it should be node to segments mapping.

Reply:
@letian-jiang nice catch. Will fix in next pr.